File: scripts/remove-bg.js
import sharp from 'sharp';

const inputPath = '/vercel/share/v0-project/public/images/hero-black-coat.jpg';
const outputPath = '/vercel/share/v0-project/public/images/hero-black-coat-cutout.png';

async function removeBg() {
  try {
    const image = sharp(inputPath);
    const metadata = await image.metadata();
    const { data, info } = await image
      .ensureAlpha()
      .raw()
      .toBuffer({ resolveWithObject: true });

    const { width, height, channels } = info;
    console.log(`Image: ${width}x${height}, channels: ${channels}`);

    const output = Buffer.from(data);

    const threshold = 228;
    const edgeThreshold = 215;

    const isBackground = new Uint8Array(width * height);
    const queue = [];

    const idx = (x, y) => (y * width + x) * channels;
    const pIdx = (x, y) => y * width + x;

    const isLight = (x, y, thresh) => {
      const i = idx(x, y);
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      const max = Math.max(r, g, b);
      const min = Math.min(r, g, b);
      const saturation = max === 0 ? 0 : (max - min) / max;
      return r >= thresh && g >= thresh && b >= thresh && saturation < 0.12;
    };

    // Seed from all edges
    for (let x = 0; x < width; x++) {
      if (isLight(x, 0, edgeThreshold)) {
        queue.push([x, 0]);
        isBackground[pIdx(x, 0)] = 1;
      }
      if (isLight(x, height - 1, edgeThreshold)) {
        queue.push([x, height - 1]);
        isBackground[pIdx(x, height - 1)] = 1;
      }
    }
    for (let y = 0; y < height; y++) {
      if (isLight(0, y, edgeThreshold)) {
        queue.push([0, y]);
        isBackground[pIdx(0, y)] = 1;
      }
      if (isLight(width - 1, y, edgeThreshold)) {
        queue.push([width - 1, y]);
        isBackground[pIdx(width - 1, y)] = 1;
      }
    }

    console.log(`Starting flood fill with ${queue.length} edge seeds...`);

    // BFS flood fill
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];

    while (queue.length > 0) {
      const [cx, cy] = queue.shift();
      for (let d = 0; d < 4; d++) {
        const nx = cx + dx[d];
        const ny = cy + dy[d];
        if (nx < 0 || nx >= width || ny < 0 || ny >= height) continue;
        if (isBackground[pIdx(nx, ny)]) continue;
        if (isLight(nx, ny, threshold)) {
          isBackground[pIdx(nx, ny)] = 1;
          queue.push([nx, ny]);
        }
      }
    }

    let bgCount = 0;

    // Shadow zone: bottom ~18% of image
    const shadowStartY = Math.round(height * 0.82);

    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        if (isBackground[pIdx(x, y)]) {
          const i = idx(x, y);

          if (y >= shadowStartY) {
            const r = data[i], g = data[i + 1], b = data[i + 2];
            const lum = (0.299 * r + 0.587 * g + 0.114 * b);

            if (lum < 245) {
              // Shadow pixel — preserve partial alpha
              const shadowStrength = 1.0 - (lum / 255.0);
              output[i + 3] = Math.round(Math.min(shadowStrength * 4.0, 1.0) * 180);
            } else {
              output[i + 3] = 0;
            }
          } else {
            output[i + 3] = 0;
          }
          bgCount++;
        }
      }
    }

    console.log(`Made ${bgCount} pixels transparent (${((bgCount / (width * height)) * 100).toFixed(1)}%)`);

    // Feather edges for softer transition
    const feathered = Buffer.from(output);
    const featherRadius = 2;

    for (let y = featherRadius; y < height - featherRadius; y++) {
      for (let x = featherRadius; x < width - featherRadius; x++) {
        const pi = pIdx(x, y);
        const i = idx(x, y);

        if (isBackground[pi]) continue;

        let hasBgNeighbor = false;
        for (let dy2 = -featherRadius; dy2 <= featherRadius && !hasBgNeighbor; dy2++) {
          for (let dx2 = -featherRadius; dx2 <= featherRadius && !hasBgNeighbor; dx2++) {
            if (dx2 === 0 && dy2 === 0) continue;
            if (isBackground[pIdx(x + dx2, y + dy2)]) hasBgNeighbor = true;
          }
        }

        if (hasBgNeighbor) {
          let bgNeighbors = 0;
          let totalNeighbors = 0;
          for (let dy2 = -featherRadius; dy2 <= featherRadius; dy2++) {
            for (let dx2 = -featherRadius; dx2 <= featherRadius; dx2++) {
              if (dx2 === 0 && dy2 === 0) continue;
              totalNeighbors++;
              if (isBackground[pIdx(x + dx2, y + dy2)]) bgNeighbors++;
            }
          }
          const ratio = bgNeighbors / totalNeighbors;
          if (ratio > 0.3) {
            feathered[i + 3] = Math.round(feathered[i + 3] * (1 - ratio * 0.6));
          }
        }
      }
    }

    await sharp(feathered, {
      raw: { width, height, channels: 4 }
    })
      .png({ compressionLevel: 9 })
      .toFile(outputPath);

    console.log(`Saved transparent PNG to ${outputPath}`);
  } catch (error) {
    console.error('Error:', error);
  }
}

removeBg();

File: scripts/remove-bg.py
"""
Remove the white/light background from the hero black coat image,
preserving the subject and the subtle ground shadow beneath the feet.

Approach:
  1. Convert to RGBA.
  2. For each pixel, calculate how close it is to pure white.
  3. Pixels very close to white become fully transparent.
  4. Pixels in the mid-range (shadow zone) get partial transparency,
     preserving the soft ground shadow as a semi-transparent overlay.
  5. Subject pixels (dark clothing, skin, hair) stay fully opaque.
"""

from PIL import Image
import numpy as np

# Load original
img = Image.open("/vercel/share/v0-project/public/images/hero-black-coat.jpg").convert("RGBA")
data = np.array(img, dtype=np.float64)

r, g, b = data[:, :, 0], data[:, :, 1], data[:, :, 2]

# Calculate "whiteness" — how close each pixel is to pure white (255,255,255).
# Using luminance-weighted distance for perceptual accuracy.
luminance = 0.299 * r + 0.587 * g + 0.114 * b

# Also check saturation — truly white/grey pixels have low saturation
max_rgb = np.maximum(np.maximum(r, g), b)
min_rgb = np.minimum(np.minimum(r, g), b)
saturation = (max_rgb - min_rgb) / (max_rgb + 1e-6)

# Whiteness score: high luminance + low saturation = background
whiteness = luminance / 255.0  # 0..1

# Define thresholds:
#   - Above high_thresh: fully transparent (pure background)
#   - Between low_thresh and high_thresh: partial transparency (shadow zone)
#   - Below low_thresh: fully opaque (subject)
high_thresh = 0.92  # Pixels brighter than this → transparent
low_thresh = 0.72   # Pixels darker than this → opaque

# Build alpha channel
alpha = np.ones_like(luminance) * 255.0

# Pure background → transparent
bg_mask = (whiteness > high_thresh) & (saturation < 0.08)
alpha[bg_mask] = 0.0

# Shadow / transition zone → partial transparency
shadow_mask = (whiteness > low_thresh) & (whiteness <= high_thresh) & (saturation < 0.12)
# Linear ramp from opaque at low_thresh to transparent at high_thresh
shadow_alpha = 1.0 - (whiteness[shadow_mask] - low_thresh) / (high_thresh - low_thresh)
alpha[shadow_mask] = shadow_alpha * 255.0

# Edge feathering: apply a very slight gaussian blur to the alpha channel
# to soften hard edges (staircase artifacts from thresholding)
from PIL import ImageFilter

alpha_img = Image.fromarray(alpha.astype(np.uint8), mode='L')
alpha_img = alpha_img.filter(ImageFilter.GaussianBlur(radius=1.2))

# Compose final image
data[:, :, 3] = np.array(alpha_img, dtype=np.float64)
result = Image.fromarray(data.astype(np.uint8), mode='RGBA')

# Save as PNG with transparency
output_path = "/vercel/share/v0-project/public/images/hero-black-coat-cutout.png"
result.save(output_path, "PNG", optimize=True)

print(f"[v0] Saved transparent PNG to {output_path}")
print(f"[v0] Image size: {result.size}")
print(f"[v0] Transparent pixels: {np.sum(np.array(alpha_img) == 0)} / {alpha.size}")
